torchtune/training/quantization.py

```diff
@@ -127,6 +152,80 @@ def quantize(self, model):
 ] = enable_8da4w_fake_quant_module_swap
 
 
+class Int8MixedPrecisionTrainingQuantizer:
+    """Apply INT8 mixed-precision training. This only affects weights of ``nn.Linear``
+    modules. During training, weights and activations are dynamically quantized to INT8
+    to utilize fast matrix multiplication with INT8 tensor cores. This is also done in
+    the backward pass.
+
+    The expected end2end speedup is 40% on a single A100 and 70% on a single 4090, with
+    minimal accuracy loss. If convergence is an issue, please refer to torchao
+    documentation below.
+
+    For more details, as well as details about arguments of this quantizer, please refer to
+    https://github.com/pytorch/ao/tree/main/torchao/prototype/quantized_training#int8-mixed-precision
+
+    Args:
+        output (bool): whether to apply INT8 mixed-precision for calculating output.
+        grad_input (bool): whether to apply INT8 mixed-precision for calculating grad_input.
+        grad_weight (bool): whether to apply INT8 mixed-precision for calculating grad_weight.
+
+    Raises:
+        RuntimeError: If runtime requirements for INT8 mixed-precision training are not met.
+
+    NOTE: Due to the limitations of the current implementation, the following
+    requirements must be satisfied to enjoy the expected speedup:
+
+    1. Must use ``torch.compile()`` (set ``compile=True``).
+    2. Inputs to the model must not be too dynamic. For example, when input tokens
+    length changes for every batch, you won't see the expected speedup.
```

Comment:
Should we warn in the recipe about one or both of these? (The compile one at least seems fairly important). Also just out of curiosity, what exactly does "too dynamic" mean? I know you mention in the summary about triggering autotune, under what conditions does this occur?

Reply:
The autotune is triggered every time input shapes and layouts change https://github.com/pytorch/ao/blob/9ce7ebbadc2433dc2935b77b4ada0e4d9e54fcb7/torchao/prototype/quantized_training/int8_mm.py#L52

Input sequence length changes for every batch will be considered "too dynamic".
Input sequence length only has a few possible values, like 256, 512, 1024, is dynamic, but still ok, since we can just autotune on these fixed set of values.

Reply:
Can we maybe use mark_dynamic to avoid recompilation if dataset.packed is false?

Reply:
The autotune logic is done by Triton (since this is a hand-written Triton kernel). `mark_dynamic` is for torch.compile(). It won't do anything in this case.

Reply:
n00b question, beyond this PR's scope, do you know if torch.compile can be used on a triton module? And if so, then maybe mark_dynamic be used?

Reply:
From what I understand, torch.compile() will treat a triton kernel as a "black box". Even if it can "see inside" the triton kernel, it can't possibly bypass/change the autotune logic. If we want the kernel to work with more "dynamicism", the solution would be to write custom autotune logic (i.e. when to trigger a re-tuning), which is not exactly simple.